lib/topological_inventory/satellite/receptor/client/response_worker.rb

```diff
@@ -0,0 +1,138 @@
+      def register_message(msg_id, receiver, response_callback: :response_received, timeout_callback: :response_timeout)
+        registered_messages[msg_id] = {:receiver => receiver, :response_callback => response_callback, :timeout_callback => timeout_callback, :registered_at => Time.now.utc}
+      end
+
+      private
+
+      attr_accessor :config, :lock, :logger, :registered_messages, :workers
+      attr_writer :started
+
+      def listen
+        # Open a connection to the messaging service
+        client = ManageIQ::Messaging::Client.open(default_messaging_opts)
+
+        logger.info("Receptor Response worker started...")
+        client.subscribe_topic(queue_opts) do |message|
+          process_message(message)
+        end
+      ensure
+        client&.close
+      end
+
+      def process_message(message)
+        response = JSON.parse(message.payload)
+        if response['code'] == 0
+          message_id = response['in_response_to']
+          # message_type: "response" (with data) or
+          #               "eof"(without data)
+          message_type = response['message_type']
+
+          if message_id
+            if (callbacks = registered_messages[message_id]).present?
+              registered_messages.delete(message_id) if message_type == 'eof'
+              # Callback to sender
+              callbacks[:receiver].send(callbacks[:response_callback], message_id, message_type, response['payload'])
+            end
+          else
+            raise "Message id (in_response_to) not received! #{response}"
+          end
+        else
+          logger.error("Receptor_satellite:health_check directive failed in receptor_client node #{response['sender']}")
+        end
+      rescue JSON::ParserError => e
+        logger.error("Failed to parse Kafka response (#{e.message})\n#{message.payload}")
+      rescue StandardError => e
+        logger.error("#{e}\n#{e.backtrace.join("\n")}")
+      end
+
+      def check_timeouts(threshold = 5.minutes)
+        while started.value do
+          expired = []
+          #
+          # STEP 1 Collect expired messages
+          #
+          registered_messages.each_pair do |message_id, callbacks|
+            if callbacks[:registered_at] < Time.now.utc - threshold
+              expired << message_id
+            end
+          end
+
+          #
+          # STEP 2 Remove expired messages, send timeout callbacks
+          #
+          expired.each do |message_id|
+            callbacks = registered_messages.delete(message_id)
+            if callbacks[:receiver].respond_to?(callbacks[:timeout_callback])
+              callbacks[:receiver].send(callbacks[:timeout_callback], message_id)
+            end
+          end
+
+          sleep(10)
+        end
+      end
+
+      def queue_opts
+        {
+          # :max_bytes   => 50_000,
+          # :persist_ref => "topological-inventory-receptor-responses", # groups subscribers (message is consumed by only one of the subscribers)
```

Comment:
@agrare do you know if this is correct settings?
1) `max_bytes`: If there is no `max_bytes`, then it's unlimited on this side?
2) `persist_ref`
Usually we need to read the message by exactly one subscriber, but there we need that all subscribers (if operations worker will be scaled) receives the message from kafka. 
We're using options like `group_ref` etc. in other repos  but I didn't find it manageiq-messaging

Reply:
@slemrmartin there might be some kafka default max_bytes but I'm not sure exactly, we'd need to check the ruby-kafka gem or the kafka broker docs specifically
If you don't pass a persist_ref then all subscribers in the group will be delivered the message